src/main/java/org/elasticsearch/common/rounding/TimeZoneRounding.java

```diff
@@ -147,21 +122,36 @@ public byte id() {
 
         @Override
         public long roundKey(long utcMillis) {
-            long offset = preTz.getOffset(utcMillis);
-            long time = utcMillis + offset;
-            return field.roundFloor(time) - offset;
+            long timeLocal = utcMillis;
+            if (!isUtc) {
+                timeLocal = timeZone.convertUTCToLocal(utcMillis);
+            }
+            long rounded = field.roundFloor(timeLocal);
+            if (!isUtc) {
+                return timeZone.convertLocalToUTC(rounded, true, utcMillis);
```

Comment:
@jpountz Doing the conversion back from local time to UTC here makes more sense to me, but I'm still wondering about performance maybe beeing better if this would be in valueForKey(). On the other hand that would implicitly create a dependency about how the two methods should be used. Should I make performance tests here?